lib/application.js

```diff
@@ -42,9 +41,14 @@ module.exports = class Application extends Emitter {
     this.middleware = [];
     this.subdomainOffset = 2;
     this.env = process.env.NODE_ENV || 'development';
-    this.context = Object.create(context);
-    this.request = Object.create(request);
-    this.response = Object.create(response);
+
+    this.Context = class AppContext extends Context {};
+    this.Request = class AppRequest extends Request {};
+    this.Response = class AppResponse extends Response {};
+
+    this.context = this.Context.prototype;
+    this.request = this.Request.prototype;
+    this.response = this.Response.prototype;
   }
```

Comment:
To preserve original behavior.
On Sun, Dec 20, 2015 at 15:22 Louis DeScioli <notifications@github.com>
wrote:

> In lib/application.js
> <https://github.com/koajs/koa/pull/605#discussion_r48108362>:
>
> > @@ -42,9 +41,14 @@ module.exports = class Application extends Emitter {
> >      this.middleware = [];
> >      this.subdomainOffset = 2;
> >      this.env = process.env.NODE_ENV || 'development';
> > -    this.context = Object.create(context);
> > -    this.request = Object.create(request);
> > -    this.response = Object.create(response);
> > +
> > +    this.Context = class AppContext extends Context {};
> > +    this.Request = class AppRequest extends Request {};
> > +    this.Response = class AppResponse extends Response {};
> > +
> > +    this.context = this.Context.prototype;
> > +    this.request = this.Request.prototype;
> > +    this.response = this.Response.prototype;
> >    }
>
> What's your thinking behind subclassing Context, Request, and Response
> here? And then saving the original class reference anyway? Why not this.Context
> = Context etc, and drop lines 49-51?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/koajs/koa/pull/605/files#r48108362>.
>